content/blog/2020-08-05-cml-self-hosted-runners-on-demand-with-gpus.md

```diff
@@ -0,0 +1,198 @@
+runners.
+
+In our recently-released project, Continuous Machine Learning (CML), we’re
+creating ways to adapt tools you already know (like GitHub Actions & GitLab CI)
+to ML. To specifically address this hardware issue, we’ve introduced a tool
+called cloud-runner. It acts as a thin wrapper over GitLab and GitHub runners,
+adding some extra capabilities to help you use self-hosted runners for ML
+workflows.
+
+With CML docker images launching your own self-hosted runner is very easy. They
+comes in two flavours, depending if you need python 2 or 3:
+
+- dvcorg/cml:latest
+- dvcorg/cml-py3:latest
+
+These images have CML and DVC preinstalled (among other perks), plus CUDA
+drivers and Python 2.7. The cml-py3 image comes with Python 3.X as well. That's
+all- you can clone these images and add your own dependencies to better mimic
+your own production environment.
+
+Using these runners will not only allow us to access special computing
+capabilities as stated, will also help us with:
+
+1.  Optimize resource usage. Using the CI/CD with our GPU hardware you can just
+    focus on experimenting, not turning machines on and off. When your workflow
+    automatically provisions runners, you maximize time and resources by only
+    using compute time when you’re training.
+
+2.  Get out of dependency hell. As data scientists, we tend to install packages
+    (on top of packages, on top of packages…) while we‘re experimenting with
+    models. In ML in particular, we can be dependent on drivers AND libraries,
+    and sometimes precise versions of them (CUDA and TensorFlow, anyone?). This
+    complexity makes it difficult to maintain a reproducible training
+    environment that’s easy to rebuild on a moment’s notice and compare to your
+    eventual production environment. Using a CI system that codifies
+
+3.  Always know what you did. One of the biggest technical debts in the ML space
+    is reproducibility. A few weeks post-experiment, we often discover that
+    trying to put your model back in shape is a pain. Looking at our repo, it’s
+    not obvious what data or training infrastructure or dependencies went into a
+    given result. Having a record that an experiment happened isn’t the same as
+    being able to recreate it. But just moving your ML experiments into a CI/CD
+    system constrains the environment and hardware used for your experiment-the
+    workflow is perfectly isolated by the containerised runner and perfectly
+    reproducible by yourself in the future or even other peers.
+
+Benefits of using CML cloud runner:
+
+- Easy to install or use. Working the same way for both vendors.
+- Maximise the use of your resources using the CI/CD.
+- Reproducibility of your ML experiments. Resolving dependencies issues and
+  tainted environments.
+
+## Hands on GPU Self-hosted runners 101
+
+### 1) Install nvidia drivers and nvidia-docker in your machine (ubuntu 18.04)
+
+```sh
```

Comment:
let's use `dvc` and start each line with `$ `

Reply:
@shcheklein sorry I don't understand what do you mean with ```dvc```

Reply:
ah, sorry

I mean that we have a custom highlighter that work better with console. You can enable it by changing `sh` to `dvc`. (it should be renamed into `console` or something eventually).